Add interop with Anvil's ContributesBinding.boundType

### Description

```kotlin
@DependencyGraph(AppScope::class)
interface MetroAppGraph {
  fun foo(): Foo
}

interface Foo
interface Bar

@com.squareup.anvil.annotations.ContributesBinding(AppScope::class, boundType = Foo::class)
class FooBinding @Inject constructor() : Foo, Bar
```
However switching to use the metro ContributesBinding works:
```kotlin
@dev.zacsweers.metro.ContributesBinding(AppScope::class, boundType = BoundType<Foo>())
```

### Self-contained Reproducer

Bottom of compilation error:
```
Caused by: java.lang.IllegalArgumentException: List has more than one element.
        at kotlin.collections.CollectionsKt___CollectionsKt.single(_Collections.kt:618)
        at dev.zacsweers.metro.compiler.fir.generators.ContributionsFirGenerator.buildBindingProperty(ContributionsFirGenerator.kt:232)
        at dev.zacsweers.metro.compiler.fir.generators.ContributionsFirGenerator.generateProperties(ContributionsFirGenerator.kt:217)
```

### Metro version

1.0.0-SNAPSHOT

### Context

_No response_

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/checkers/AggregationChecker.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/checkers/AggregationChecker.kt
@@ -203,7 +203,7 @@ internal object AggregationChecker : FirClassChecker(MppCheckerKind.Common) {
     val supertypesExcludingAny = declaration.superTypeRefs.filterNot { it.coneType.isAny }
     val hasSupertypes = supertypesExcludingAny.isNotEmpty()
 
-    val explicitBoundType = annotation.resolvedBoundType()
+    val explicitBoundType = annotation.resolvedBoundType(session)
 
     val typeKey =
       if (explicitBoundType != null) {

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
@@ -30,8 +30,10 @@ import org.jetbrains.kotlin.fir.declarations.FirDeclarationOrigin
 import org.jetbrains.kotlin.fir.declarations.FirMemberDeclaration
 import org.jetbrains.kotlin.fir.declarations.FirProperty
 import org.jetbrains.kotlin.fir.declarations.FirTypeParameterRef
+import org.jetbrains.kotlin.fir.declarations.evaluateAs
 import org.jetbrains.kotlin.fir.declarations.findArgumentByName
 import org.jetbrains.kotlin.fir.declarations.getDeprecationsProvider
+import org.jetbrains.kotlin.fir.declarations.getTargetType
 import org.jetbrains.kotlin.fir.declarations.hasAnnotation
 import org.jetbrains.kotlin.fir.declarations.origin
 import org.jetbrains.kotlin.fir.declarations.primaryConstructorIfAny
@@ -91,6 +93,7 @@ import org.jetbrains.kotlin.fir.types.ConeClassLikeType
 import org.jetbrains.kotlin.fir.types.ConeKotlinType
 import org.jetbrains.kotlin.fir.types.ConeTypeProjection
 import org.jetbrains.kotlin.fir.types.FirTypeProjectionWithVariance
+import org.jetbrains.kotlin.fir.types.FirTypeRef
 import org.jetbrains.kotlin.fir.types.builder.buildResolvedTypeRef
 import org.jetbrains.kotlin.fir.types.classId
 import org.jetbrains.kotlin.fir.types.coneTypeOrNull
@@ -788,12 +791,27 @@ internal fun FirAnnotation.replacesArgument() =
 
 internal fun FirAnnotation.boundTypeArgument() = annotationArgument("boundType".asName(), index = 2)
 
-internal fun FirAnnotation.resolvedBoundType() =
-  boundTypeArgument()
-    ?.typeArguments
-    ?.get(0)
-    ?.expectAsOrNull<FirTypeProjectionWithVariance>()
-    ?.typeRef
+internal fun FirAnnotation.resolvedBoundType(session: FirSession): FirTypeRef? {
+  // Return a BoundType defined using metro api's
+  boundTypeArgument()?.let { boundType ->
+    return boundType.typeArguments[0].expectAsOrNull<FirTypeProjectionWithVariance>()?.typeRef
+  }
+  // Return a boundType defined using anvil KClass
+  return kClassBoundTypeArgument(session)
+}
+
+internal fun FirAnnotation.kClassBoundTypeArgument(session: FirSession): FirTypeRef? {
+  return getAnnotationKClassArgument("boundType".asName(), session)?.toFirResolvedTypeRef()
+}
+
+internal fun FirAnnotation.getAnnotationKClassArgument(
+  name: Name,
+  session: FirSession,
+): ConeKotlinType? {
+  val argument = findArgumentByName(name) ?: return null
+  val getClassCall = argument.evaluateAs<FirGetClassCall>(session) ?: return null
+  return getClassCall.getTargetType()
+}
 
 internal fun FirAnnotation.resolvedScopeClassId() = scopeArgument()?.resolvedClassId()
 

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributionsFirGenerator.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/generators/ContributionsFirGenerator.kt
@@ -12,6 +12,7 @@ import dev.zacsweers.metro.compiler.fir.buildSimpleAnnotation
 import dev.zacsweers.metro.compiler.fir.classIds
 import dev.zacsweers.metro.compiler.fir.hasOrigin
 import dev.zacsweers.metro.compiler.fir.isAnnotatedWithAny
+import dev.zacsweers.metro.compiler.fir.kClassBoundTypeArgument
 import dev.zacsweers.metro.compiler.fir.mapKeyAnnotation
 import dev.zacsweers.metro.compiler.fir.markAsDeprecatedHidden
 import dev.zacsweers.metro.compiler.fir.metroFirBuiltIns
@@ -192,12 +193,16 @@ internal class ContributionsFirGenerator(session: FirSession) :
   }
 
   private fun FirAnnotation.boundTypeOrNull(): FirTypeRef? {
-    return argumentAsOrNull<FirFunctionCall>("boundType".asName(), 2)
-      ?.typeArguments
-      ?.getOrNull(0)
-      ?.expectAsOrNull<FirTypeProjectionWithVariance>()
-      ?.typeRef
-      ?.takeUnless { it == session.builtinTypes.nothingType }
+    // Return a BoundType defined using metro api's
+    argumentAsOrNull<FirFunctionCall>("boundType".asName(), 2)?.let { boundType ->
+      return boundType.typeArguments
+        .getOrNull(0)
+        ?.expectAsOrNull<FirTypeProjectionWithVariance>()
+        ?.typeRef
+        ?.takeUnless { it == session.builtinTypes.nothingType }
+    }
+    // Return a boundType defined using anvil KClass
+    return kClassBoundTypeArgument(session)
   }
 
   override fun generateProperties(

--- a/compiler/src/test/kotlin/dev/zacsweers/metro/compiler/fir/AggregationTest.kt
+++ b/compiler/src/test/kotlin/dev/zacsweers/metro/compiler/fir/AggregationTest.kt
@@ -14,6 +14,7 @@ import dev.zacsweers.metro.compiler.generatedMetroGraphClass
 import kotlin.reflect.KClass
 import kotlin.test.Test
 import kotlin.test.assertNotNull
+import org.jetbrains.kotlin.name.ClassId
 import org.junit.Ignore
 
 class AggregationTest : MetroCompilerTest() {
@@ -1071,6 +1072,52 @@ class AggregationTest : MetroCompilerTest() {
       val contributedInterface2 = graph.callProperty<Any>("contributedInterface2")
       assertThat(contributedInterface2).isNotNull()
       assertThat(contributedInterface2.javaClass.name).isEqualTo("test.Impl")
+    }
+  }
+
+  @Test
+  fun `repeated ContributesBinding annotations with different scopes and same bound types are ok2`() {
+    compile(
+      source(
+        packageName = "com.squareup.anvil.annotations",
+        source =
+          """
+            annotation class ContributesBinding(
+              val scope: KClass<*>,
+              val boundType: KClass<*> = Unit::class,
+            )
+            """
+            .trimIndent(),
+      ),
+      source(
+        """
+          import com.squareup.anvil.annotations.ContributesBinding
+          import dev.zacsweers.metro.Inject
+
+          interface ContributedInterface
+          interface SecondInterface
+
+          @ContributesBinding(AppScope::class, boundType = ContributedInterface::class)
+          @Inject
+          class Impl : ContributedInterface, SecondInterface
+
+          @DependencyGraph(scope = AppScope::class)
+          interface ExampleGraph {
+            val contributedInterface: ContributedInterface
+          }
+        """
+          .trimIndent()
+      ),
+      options =
+        metroOptions.copy(
+          customContributesBindingAnnotations =
+            setOf(ClassId.fromString("com/squareup/anvil/annotations/ContributesBinding"))
+        ),
+    ) {
+      val graph = ExampleGraph.generatedMetroGraphClass().createGraphWithNoArgs()
+      val contributedInterface = graph.callProperty<Any>("contributedInterface")
+      assertThat(contributedInterface).isNotNull()
+      assertThat(contributedInterface.javaClass.name).isEqualTo("test.Impl")
     }
   }
 

--- a/docs/interop.md
+++ b/docs/interop.md
@@ -49,7 +49,9 @@ metro {
 
 `@DependencyGraph` is replaceable but your mileage may vary if you use Anvil or modules, since Metro’s annotation unifies Anvil’s `@MergeComponent` functionality and doesn’t support modules.
 
-Similarly, `@ContributesBinding` is replaceable but there are not direct analogues for Anvil’s `@ContributesMultibinding` or kotlin-inject-anvil’s `@ContributesBinding(multibinding = …)` as these annotations are implemented as `@ContributesInto*` annotations in Metro. Also - `boundType` in metro uses a more flexible mechanism to support generics.
+Similarly, `@ContributesBinding` is replaceable but there are not direct analogues for Anvil’s `@ContributesMultibinding` or kotlin-inject-anvil’s `@ContributesBinding(multibinding = …)` as these annotations are implemented as `@ContributesInto*` annotations in Metro.
+
+`boundType` in Metro uses a more flexible mechanism to support generics, but interop with Anvil's `boundType: KClass<*>` property is supported.
 
 ## Components
 

--- a/samples/interop/customAnnotations-dagger/src/test/kotlin/dev/zacsweers/metro/sample/AnvilTest.kt
+++ b/samples/interop/customAnnotations-dagger/src/test/kotlin/dev/zacsweers/metro/sample/AnvilTest.kt
@@ -22,6 +22,7 @@ class AnvilTest {
 
     val message: String
     val baseClass: BaseClass
+    val baseClass2: BaseClass2
 
     @MergeComponent.Factory
     interface Factory {
@@ -36,6 +37,13 @@ class AnvilTest {
   @ContributesBinding(AppScope::class)
   class Impl @Inject constructor(override val message: String) : BaseClass
 
+  interface BaseClass2 {
+    val message: String
+  }
+
+  @ContributesBinding(AppScope::class, boundType = BaseClass2::class)
+  class Impl2 @Inject constructor(override val message: String) : BaseClass2
+
   @ContributesTo(AppScope::class) interface ContributedInterface
 
   @Test
@@ -46,5 +54,6 @@ class AnvilTest {
     assertTrue(component is ContributedInterface)
 
     assertEquals("Hello, world!", component.baseClass.message)
+    assertEquals("Hello, world!", component.baseClass2.message)
   }
 }

